### PR TITLE
Land changes that require Kotlin 1.6

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Comment by swankjesse** [comment]
# Do not merge

Waiting on downstream repos to be ready for Kotlin 1.6.

File: zipline-kotlin-plugin/tests/src/test/java/app/cash/zipline/kotlin/ZiplineTestInternals.java [hunk 1]
```diff
@@ -68,113 +68,178 @@
...
-          }
-        };
-      }
-    });
+    return endpoint.get(name, GenericEchoServiceAdapter.INSTANCE);
   }
 
   /** Simulate generated code for inbound calls. */
   public static void setGenericEchoService(
       Endpoint endpoint, String name, GenericEchoService<String> echoService) {
```
Comment (veyndan) [inline zipline-kotlin-plugin/tests/src/test/java/app/cash/zipline/kotlin/ZiplineTestInternals.java:87]: What’s happening here?

File: zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt [hunk 2]
```diff
@@ -218,31 +146,35 @@ class ZiplineKotlinPluginTest {
...
       )
     )
     assertEquals(KotlinCompilation.ExitCode.COMPILATION_ERROR, result.exitCode, result.messages)
     assertThat(result.messages)
-      .contains("(6, 19): The type argument to Zipline.get() must be an interface type")
+      .contains("(9, 19): The type argument to Zipline.take() must be an interface type")
   }
 
   @Test
+  @Ignore("generics are broken after the ZiplineService refactoring")
```
Comment (veyndan) [inline zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt:177]: Should this be fixed before merging (or at least have a GH issue regarding it)?

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -53,7 +53,7 @@ jobs:
           cmake --build build/jni/${{ matrix.arch }}/ --verbose
           cp -v build/jni/${{ matrix.arch }}/libquickjs.* zipline/src/jvmMain/resources/jni/${{ matrix.arch }}/
 
-      - run: ./gradlew build
+      - run: ./gradlew build --stacktrace
```
Comment (kevincianfarini) [inline .github/workflows/build.yaml:56]: We should probably remove this now that debugging a build is presumably done, right?

File: zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt [hunk 4]
```diff
@@ -317,15 +338,15 @@ internal class AdapterGenerator(
...
       +irReturn(
-        irCallFunctionBody(bridgedInterface, callFunction, serviceProperty)
+        irCallFunctionBody(bridgedInterface, callFunction, serviceProperty, supportedFunctionNamesProperty)
       )
     }
     callSuspendingFunction.irFunctionBody(
       context = pluginContext,
       scopeOwnerSymbol = callSuspendingFunction.symbol,
     ) {
       +irReturn(
```
Comment (kevincianfarini) [inline zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt:348]: drive by, uncontextual question: Why are we using the unary `+` operator here?

Comment (JakeWharton) [reply zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt:348]: It's the Kotlin compiler plugin API

Comment (JakeWharton) [reply zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt:348]: And it's how you add new IR to a block

**Review by JakeWharton — APPROVED** [review]
Hard to review without a rebase. Mostly just stamping.

File: zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt (hunk 4 above)
Comment (kevincianfarini) [reply zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt:348]: Oh gross. Jetbrains really loves leveraging operator overloads.

Comment (JakeWharton) [reply zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt:348]: It isn't great, but you get used to it. Definitely wouldn't want something parenthesized like `add(irReturn(..))` but maybe `this += irReturn(..)` would be a happy medium.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (oldergod) [reply .github/workflows/build.yaml:56]: It's valuable to keep it around. Doesn't pollute when green, and insightful when red. No?

File: zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt (hunk 2 above)
Comment (swankjesse) [reply zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt:177]: Good call on a GitHub issue. I think the ultimate solution might be just dropping this behavior.

Comment (swankjesse) [reply zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt:177]: https://github.com/cashapp/zipline/issues/420

File: .github/workflows/build.yaml (hunk 3 above)
Comment (swankjesse) [reply .github/workflows/build.yaml:56]: Yeah, I like it.

**Pull request merged**